chore: bump postgresql and gitlab

ableuler opened the pull request:
This PR bumps postgresql from 9.6 to 11 and GitLab from 11.9.11 to 13.10.4 and adds instructions and helper k8s manifests on how to do this.Tested on a personal dev deployment and on a deployment on a personal iac cluster which contains gitlab.

/deploy

olevski commented on helm-chart/utils/postgres_migrations/version_upgrades/README.md:
````diff
@@ -0,0 +1,48 @@
+- Modify the example manifest files in this folder according to your needs. You can use the followig command to figure out the total amount of data currently stored in your database. This will give you a fair estimate of the size of your data dump.
+
+```sql
+SELECT pg_database.datname as "database_name", pg_database_size(pg_database.datname)/1024/1024 AS size_in_mb FROM pg_database ORDER by size_in_mb DESC;
````
We can replace this with a clear way of how to access the postgres instance. Or at least add it as a suggestion.
```bash
```bash
kubectl -n renku exec -ti renku-postgresql-0 -- bash
PGUSER=$POSTGRES_USER PGPASSWORD=$POSTGRES_PASSWORD psql -c 'SELECT pg_database.datname as "database_name", pg_database_size(pg_database.datname)/1024/1024 AS size_in_mb FROM pg_database ORDER by size_in_mb DESC;'
```
```

ableuler replied:
done

pameladelgado commented on helm-chart/README.rst:
```diff
@@ -64,6 +64,46 @@ in the `values changelog file <https://github.com/SwissDataScienceCenter/renku/b
+
+1. Upgrading postgresql
++++++++++++++++++++++++
+Follow [these instructions](https://github.com/SwissDataScienceCenter/renku/tree/master/helm-chart/utils/postgres_migrations/version_upgrades/README.md)
```
```suggestion
If PostgreSQL was deployed as part of Renku, please follow [these instructions](https://github.com/SwissDataScienceCenter/renku/tree/master/helm-chart/utils/postgres_migrations/version_upgrades/README.md)
```

pameladelgado commented on helm-chart/README.rst:
```diff
@@ -64,6 +64,46 @@ in the `values changelog file <https://github.com/SwissDataScienceCenter/renku/b
 
+Upgrading to 0.8.0
+******************
+We bump the posgresql version from 9.6 to 11 and the GitLab major version from 11 to 13.
```
```suggestion
We bump the PostgreSQL version from 9.6 to 11 and the GitLab major version from 11 to 13.
```

pameladelgado commented on helm-chart/README.rst:
```diff
@@ -64,6 +64,46 @@ in the `values changelog file <https://github.com/SwissDataScienceCenter/renku/b
+
+2. Bump the chart version
++++++++++++++++++++++++++
+Now it's time to upgrade to the 0.8.0 version of the Renku chart. Before doing this, make sure
```
```suggestion
Now it's time to upgrade to the ``0.8.0`` version of the Renku chart. Before doing this, make sure
```

pameladelgado commented on helm-chart/README.rst:
```diff
@@ -64,6 +64,46 @@ in the `values changelog file <https://github.com/SwissDataScienceCenter/renku/b
+to pin the GitLab version by setting ``gitlab.image.tag`` in your values file. If you had not pinned
+this version explicitly before, pin it to ``11.9.11-ce.0`` which is the default version set in the Renku
+chart prior to the upgrade. Otherwise you can leave it at the previously pinned version. Then deploy the
+the new chart version through ``helm upgrade ... --version 0.8.0 ...``.
```
```suggestion
new chart version through ``helm upgrade ... --version 0.8.0 ...``.
```

pameladelgado commented on helm-chart/utils/postgres_migrations/version_upgrades/README.md:
```diff
@@ -0,0 +1,48 @@
+# Upgrading postgresql
+
+Here we describe how the version of a postgresql instance which is installed through the Renku chart can be upgraded. While minor version upgrades are unproblematic and do not need any manual intervention, major version upgrades require some care. There are two recommended ways of performing this task:
```
```suggestion
Here we describe how the version of a PostgreSQL instance which is installed through the Renku chart can be upgraded. While minor version upgrades are unproblematic and do not need any manual intervention, major version upgrades require some care. There are two recommended ways of performing this task:
```

pameladelgado commented on helm-chart/utils/postgres_migrations/version_upgrades/README.md:
```diff
@@ -0,0 +1,48 @@
+# Upgrading postgresql
```
we could add a comment to clarify that these instructions are not specific to Renku but just a guideline on how we (will) upgrade the PostgreSQL version in our deployments, otherwise people are welcome to use their own methods as well.

ableuler replied:
done

pameladelgado commented on helm-chart/utils/postgres_migrations/version_upgrades/README.md:
```diff
@@ -0,0 +1,48 @@
+
+Here we describe how the version of a postgresql instance which is installed through the Renku chart can be upgraded. While minor version upgrades are unproblematic and do not need any manual intervention, major version upgrades require some care. There are two recommended ways of performing this task:
+
+1) Dump all the data in the old instance using `pd_dumpall`, start a fresh instance using the target version and import the previous data dump.
```
```suggestion
1) Dump all the data in the old instance using `pd_dumpall`, start a fresh instance using the target version and import the previous data dump. Instructions below.
```

pameladelgado commented on helm-chart/utils/postgres_migrations/version_upgrades/README.md:
```diff
@@ -0,0 +1,48 @@
+## Preparation
+
+- Modify the example manifest files in this folder according to your needs. You can use the followig command to figure out the total amount of data currently stored in your database. This will give you a fair estimate of the size of your data dump.
+
```
Add a step: start a shell in the postgresql pod

pameladelgado commented on helm-chart/utils/postgres_migrations/version_upgrades/README.md:
````diff
@@ -0,0 +1,48 @@
+```
+
+- If you're not using dynamic volume provisioning, create a new empty PV/PVC that will contain the data of your upgraded postresql instance.
+- Remove the renku postgresql statefulSet. You might also want to scale down the services which try to connect to the postgresql instance, but it is not strictly required.
````
```suggestion
- Remove the renku PostgreSQL statefulSet. You might also want to scale down the services which try to connect to the PostgreSQL instance, but it is not strictly required.
```

pameladelgado reviewed (commented):
Thanks @ableuler !
This is looking good, I have a few minor comments.
Other than that, is there a reason why we wouldnt want to pin the minor version of PostgreSQL? e.g. `11.9` . I think using just `11` in the documentation and readmes is ok but for the image I think is better to use the latest supported minor version explicitly.

olevski commented on helm-chart/utils/postgres_migrations/version_upgrades/README.md:
```diff
@@ -0,0 +1,48 @@
+## Dumping the data
+
+- Execute `kubectl apply -f psql_dump.yaml`, wait for the `postgresql-dump` pod to become available.
+- Execute the following command inside the container:
```
```suggestion
- Execute the following command inside the `postgresql-dump` container:
```

ableuler commented:
> Other than that, is there a reason why we wouldnt want to pin the minor version of PostgreSQL?

That wasn't a conscious decision to be honest. I just carried on with what we'd been doing before (9.6 is also a major version). While I totally agree in general, postgresql seems such a mature project that I'm actually wondering if we should keep pinning only the major version and use an image pull policy "Always" would make sense?

olevski commented:
So I just wanted to cofirm that I brought up renku + gitlab in my own deployment. I started with the old postgres version and did the upgrade as discussed here. Everything worked.

Then I went through the gitlab upgrade steps and everything was in order as well.

ableuler commented on helm-chart/utils/postgres_migrations/version_upgrades/README.md:
```diff
@@ -0,0 +1,48 @@
+
+## Cleanup
+
+- Delete the temporary volume which held the dump: `kubectl delete -n <your-namespace> data-renku-postgresql-tmp`. If the chosen storage class does not have the `Reclaim Policy` set to `Delete`, you want to also delete the PV.
```
Now mentioned in `## Create the volumes`

ableuler commented on helm-chart/utils/postgres_migrations/version_upgrades/README.md:
````diff
@@ -0,0 +1,48 @@
+```
+
+- If you're not using dynamic volume provisioning, create a new empty PV/PVC that will contain the data of your upgraded postresql instance.
+- Remove the renku postgresql statefulSet. You might also want to scale down the services which try to connect to the postgresql instance, but it is not strictly required.
````
I've actually removed that line entirely. I've tried this multiple times now and I've simply removed postgres but didn't scale down any clients.

ableuler commented:
I've modified the postgres upgrade part a bit to deal with multiple comments at once. The reason I went for a naked pods rather than for k8s jobs is that we need to run a postgresql server during both dumping and loading, and this isn't a process that runs to completion. Also, I don't want this to be our regular postgresql server instance since I want to avoid that any clients can write to the db while dumping/loading. But I understand that executing the actual dumping/loading inside the pod container was a bit hacky and not well described, so I've added proper k8s jobs which connect to the respective temporary server instance and perform the dumping and loading.

ableuler commented on helm-chart/utils/postgres_migrations/version_upgrades/psql_dump.yaml:
```diff
@@ -0,0 +1,55 @@
+      storage: 20Gi ## EDIT
+---
+apiVersion: v1
+kind: Pod
```
See response below.

ableuler commented on helm-chart/utils/postgres_migrations/version_upgrades/README.md:
```diff
@@ -0,0 +1,48 @@
+
+## Dumping the data
+
+- Execute `kubectl apply -f psql_dump.yaml`, wait for the `postgresql-dump` pod to become available.
```
Mentioned more clearly in the `## Preparation` section.

pameladelgado commented on helm-chart/renku/values.yaml:
```diff
@@ -230,7 +230,7 @@ postgresql:
   image:
     repository: bitnami/postgresql
-    tag: 9.6
+    tag: 11
```
should we go for `11.11.0` (latest)?
Otherwise I'd pin this in all of our values files...

olevski commented on helm-chart/utils/postgres_migrations/version_upgrades/README.md:
```diff
@@ -0,0 +1,44 @@
+
+## Dump the data
+
+- Execute `kubectl apply -n <your-namespace> -f psql_dump.yaml` and wait until the `postgresql-dump-job` has completed. Remove the temporary resources using `kubectl delete -n <your-namespace> -f psql_dump.yaml`.
```
At this point we should mention (again) that before the `psql_dump.yaml` manifest is applied the statefulset that backs postgres should be deleted.

olevski commented:
I tested the upgrade to postgres from 9.6 to 11 with the new manifests that include all the required stuff to create the volumes, dump and load data. All works beside that one comment above about deleting the statefulset before applying the dump manifests. I should have remembered that was required. So I dont think it is a bad idea to mention it.

pameladelgado commented on helm-chart/utils/postgres_migrations/version_upgrades/volumes.yaml:
```diff
@@ -0,0 +1,60 @@
+    - ReadWriteOnce
+  resources:
+    requests:
+      storage: ## EDIT - Twice the size of all current DBs combined should be more than enough.
```
why should it be double?

ableuler replied:
I read somewhere that the dump can be bigger than the reported data usage. So doubling seemed a conservative strategy.

pameladelgado commented on helm-chart/README.rst:
```diff
@@ -64,6 +64,46 @@ in the `values changelog file <https://github.com/SwissDataScienceCenter/renku/b
+- 13.10.4-ce.0 (default in the Renku ``0.8.0`` helm chart)
+
+For each step, set the corresponding tag in your values file at ``gitlab.image.tag``, redeploy through
+helm and wait for the gitlab pod to be recreated and all migrations to finish. Repeat this procedure until
```
To me, it's faster to just replace the image by editing the deployment. Especially if there are a lot of upgrades to be done. Then I'd use the `helm upgrade` only to 'persist' the latest version.

ableuler replied:
I think that's a viable option for people with more experience, but for the general instructions I'd prefer to keep it at the level of the chart values.

pameladelgado replied:
Yes, I guess it makes sense. We should have pinned newer GitLab versions in our default values file a long time ago. At least we're soon going to change this.

pameladelgado reviewed (approved):
Works like a charm! :tada: 
I have few minor comments only, not a big deal if ignored...

pameladelgado commented on helm-chart/utils/postgres_migrations/version_upgrades/psql_load.yaml:
```diff
@@ -0,0 +1,86 @@
+      targetPort: 5432
+---
+apiVersion: batch/v1
+kind: Job
```
For what is worth, the load was correct as far as I can tell and I didn't see error messages in the logs either. It was just strange to not see a `Succeeded` state.
